corehq/form_processor/backends/sql/dbaccessors.py

```diff
@@ -507,13 +511,24 @@ def save_new_form(form):
         for operation in operations:
             operation.form = form
 
-        with get_cursor(XFormInstanceSQL) as cursor:
-            cursor.execute(
-                'SELECT form_pk FROM save_new_form_and_related_models(%s, %s, %s, %s)',
-                [form.form_id, form, unsaved_attachments, operations]
-            )
-            result = fetchone_as_namedtuple(cursor)
-            form.id = result.form_pk
+        try:
+            with transaction.atomic(using=db_name):
+                form.save(using=db_name)
+                for attachment in unsaved_attachments:
+                    if attachment.is_saved():
+                        raise XFormSaveError(
+                            'XFormAttachmentSQL {} has already been saved'.format(attachment.id)
+                        )
+                    attachment.save(using=db_name)
+
+                for operation in operations:
+                    if operation.is_saved():
+                        raise XFormSaveError(
+                            'XFormOperationSQL {} has already been saved'.format(operation.id)
+                        )
+                    operation.save(using=db_name)
```

Comment:
if what you say is true (which is sounds like it is) then saving a form isn't atomic. https://github.com/dimagi/commcare-hq/blob/br/form-save-python/corehq/form_processor/backends/sql/processor.py#L77 we could successfully save the form, then fail on the case save which leaves us in a weird state. wonder if that has happened at all